File: apps/client-macos/Tests/TimeTrackTests/IdleMonitorTests.swift
import XCTest
@testable import TimeTrack

final class IdleMonitorTests: XCTestCase {
    private final class MutableClock {
        private(set) var now: Date
        init(_ start: Date) { now = start }
        func advance(_ s: TimeInterval) { now = now.addingTimeInterval(s) }
        func read() -> Date { now }
    }
    private let t0 = Date(timeIntervalSince1970: 1_700_000_000)

    private func make(threshold: Int = 300)
        -> (IdleMonitor, FakeIdleMonitorDelegate, MutableClock) {
        let clock = MutableClock(t0)
        let monitor = IdleMonitor(thresholdSeconds: threshold, clock: clock.read)
        let delegate = FakeIdleMonitorDelegate()
        monitor.delegate = delegate
        return (monitor, delegate, clock)
    }

    func testActivateStartsTracking() {
        let (monitor, delegate, _) = make()
        monitor.activate()
        XCTAssertEqual(delegate.calls, [.start])
        XCTAssertEqual(monitor.state, .active)
    }

    func testIdleThresholdStopsAtAwayStart() {
        let (monitor, delegate, clock) = make(threshold: 300)
        monitor.activate()
        clock.advance(305)                        // 305s of no input
        monitor.tick(idleSeconds: 305)
        // awayStart = now(t0+305) - 305 = t0
        XCTAssertEqual(delegate.calls, [.start, .stop(at: t0)])
        XCTAssertEqual(monitor.state, .away(since: t0))
    }

    func testSubThresholdTickDoesNotStop() {
        let (monitor, delegate, _) = make(threshold: 300)
        monitor.activate()
        monitor.tick(idleSeconds: 120)
        XCTAssertEqual(delegate.calls, [.start], "no stop below threshold")
        XCTAssertEqual(monitor.state, .active)
    }

    func testResumeAfterIdlePromptsThenKeepBridgesAndRestarts() {
        let (monitor, delegate, clock) = make(threshold: 300)
        monitor.activate()
        clock.advance(300); monitor.tick(idleSeconds: 300)   // away since t0
        clock.advance(120); monitor.tick(idleSeconds: 5)      // input again → resume at t0+420
        XCTAssertEqual(monitor.state, .awaiting(since: t0, until: t0.addingTimeInterval(420)))
        XCTAssertEqual(delegate.calls.last, .becameAway(seconds: 420))

        monitor.resolve(.keep)
        XCTAssertEqual(delegate.calls.suffix(2), [
            .resolved(from: t0, to: t0.addingTimeInterval(420), keeping: true),
            .start,
        ])
        XCTAssertEqual(monitor.state, .active)
    }

    func testResumeThenDiscardDoesNotBridgeAndRestarts() {
        let (monitor, delegate, clock) = make(threshold: 300)
        monitor.activate()
        clock.advance(300); monitor.tick(idleSeconds: 300)
        clock.advance(60); monitor.tick(idleSeconds: 1)
        monitor.resolve(.discard)
        XCTAssertEqual(delegate.calls.suffix(2), [
            .resolved(from: t0, to: t0.addingTimeInterval(360), keeping: false),
            .start,
        ])
    }

    func testSleepStopsImmediatelyAtNow() {
        let (monitor, delegate, clock) = make()
        monitor.activate()
        clock.advance(30)
        monitor.markAway()                        // sleep/lock: away starts now, not after threshold
        XCTAssertEqual(delegate.calls, [.start, .stop(at: t0.addingTimeInterval(30))])
        XCTAssertEqual(monitor.state, .away(since: t0.addingTimeInterval(30)))
    }

    func testDeactivateWhileAwayEmitsUnresolved() {
        let (monitor, delegate, clock) = make(threshold: 300)
        monitor.activate()
        clock.advance(300); monitor.tick(idleSeconds: 300)     // away since t0
        clock.advance(90)
        monitor.deactivate()                                    // e.g. sign-out mid-away
        XCTAssertEqual(delegate.calls.last, .abandoned(from: t0, to: t0.addingTimeInterval(390)))
        XCTAssertEqual(monitor.state, .inactive)
    }
}

File: apps/client-windows/tests/NiftyTimer.Tests/IdleMonitorTests.cs
using NiftyTimer.Tracking;
using Xunit;

namespace NiftyTimer.Tests;

/// <summary>
/// The automatic-tracking decision machine (PRD §6.1/§6.4). Pure: no UI, no network, no hardware.
/// </summary>
public class IdleMonitorTests
{
    private static readonly DateTimeOffset T0 = new(2026, 8, 25, 9, 0, 0, TimeSpan.Zero);

    private sealed class Recorder : IIdleMonitorDelegate
    {
        public List<string> Calls { get; } = [];

        public List<(DateTimeOffset From, DateTimeOffset To, bool Keeping)> Resolved { get; } = [];

        public List<(DateTimeOffset From, DateTimeOffset To)> Abandoned { get; } = [];

        public List<DateTimeOffset> Stops { get; } = [];

        public List<int> AwaySeconds { get; } = [];

        public List<int> ThresholdCrossings { get; } = [];

        public void ShouldStartTracking() => Calls.Add("start");

        public void ShouldStopTracking(DateTimeOffset awayStart)
        {
            Calls.Add("stop");
            Stops.Add(awayStart);
        }

        public void DidBecomeAway(int seconds)
        {
            Calls.Add("away");
            AwaySeconds.Add(seconds);
        }

        public void DidResolveAway(DateTimeOffset awayStart, DateTimeOffset resume, bool keeping)
        {
            Calls.Add("resolve");
            Resolved.Add((awayStart, resume, keeping));
        }

        public void DidAbandonAway(DateTimeOffset awayStart, DateTimeOffset lastKnown)
        {
            Calls.Add("abandon");
            Abandoned.Add((awayStart, lastKnown));
        }

        public void DidCrossIdleThreshold(int seconds)
        {
            Calls.Add("threshold");
            ThresholdCrossings.Add(seconds);
        }
    }

    private static (IdleMonitor Monitor, Recorder Delegate) New(Func<DateTimeOffset> clock, int threshold = 300)
    {
        var recorder = new Recorder();
        var monitor = new IdleMonitor(threshold, clock) { Delegate = recorder };
        return (monitor, recorder);
    }

    [Fact]
    public void StartsInactiveAndSaysNothing()
    {
        var (monitor, recorder) = New(() => T0);

        Assert.Equal(IdleState.Inactive, monitor.State);
        Assert.Empty(recorder.Calls);
    }

    [Fact]
    public void ActivateOpensASpan()
    {
        var (monitor, recorder) = New(() => T0);

        monitor.Activate();

        Assert.Equal(IdleState.Active, monitor.State);
        Assert.Equal(["start"], recorder.Calls);
    }

    [Fact]
    public void TicksBelowTheThresholdChangeNothing()
    {
        var (monitor, recorder) = New(() => T0);
        monitor.Activate();
        recorder.Calls.Clear();

        monitor.Tick(0);
        monitor.Tick(299);

        Assert.Equal(IdleState.Active, monitor.State);
        Assert.Empty(recorder.Calls);
    }

    /// <summary>
    /// The away window starts when input actually stopped, not when we noticed. The poller runs
    /// every 15s, so "now" is always late — backdating by the reported idle seconds is what keeps
    /// the excluded window equal to the real one.
    /// </summary>
    [Fact]
    public void CrossingTheThresholdStopsAtTheBackdatedAwayStart()
    {
        var now = T0.AddMinutes(30);
        var (monitor, recorder) = New(() => now);
        monitor.Activate();

        monitor.Tick(310);

        Assert.Equal(new IdleState.Away(now.AddSeconds(-310)), monitor.State);
        Assert.Equal(now.AddSeconds(-310), Assert.Single(recorder.Stops));
        Assert.Equal(310, Assert.Single(recorder.ThresholdCrossings));
    }

    [Fact]
    public void StaysAwayWhileIdleRemainsAboveTheThreshold()
    {
        var now = T0;
        var (monitor, recorder) = New(() => now);
        monitor.Activate();
        monitor.Tick(300);
        var away = monitor.State;
        recorder.Calls.Clear();

        now = T0.AddMinutes(10);
        monitor.Tick(900);

        Assert.Equal(away, monitor.State); // the away START does not drift
        Assert.Empty(recorder.Calls);
    }

    [Fact]
    public void ReturningBelowTheThresholdAsksTheUser()
    {
        var now = T0;
        var (monitor, recorder) = New(() => now);
        monitor.Activate();
        monitor.Tick(600);
        var awayStart = now.AddSeconds(-600);

        now = T0.AddMinutes(20);
        monitor.Tick(3);

        Assert.Equal(new IdleState.Awaiting(awayStart, now), monitor.State);
        Assert.Equal((int)(now - awayStart).TotalSeconds, Assert.Single(recorder.AwaySeconds));
    }

    /// <summary>
    /// Sleep and lock are known moments, so they mark away immediately rather than waiting for the
    /// threshold — a closed lid is not five minutes of maybe.
    /// </summary>
    [Fact]
    public void MarkAwayStopsImmediatelyAtNow()
    {
        var now = T0.AddHours(1);
        var (monitor, recorder) = New(() => now);
        monitor.Activate();

        monitor.MarkAway();

        Assert.Equal(new IdleState.Away(now), monitor.State);
        Assert.Equal(now, Assert.Single(recorder.Stops));
        Assert.Empty(recorder.ThresholdCrossings); // not an inactivity crossing — no idle nudge
    }

    [Fact]
    public void MarkAwayIsIgnoredUnlessActive()
    {
        var (monitor, recorder) = New(() => T0);

        monitor.MarkAway();

        Assert.Equal(IdleState.Inactive, monitor.State);
        Assert.Empty(recorder.Calls);
    }

    [Fact]
    public void KeepBridgesTheWindowAndReopens()
    {
        var now = T0;
        var (monitor, recorder) = New(() => now);
        monitor.Activate();
        monitor.Tick(600);
        var awayStart = now.AddSeconds(-600);
        now = T0.AddMinutes(30);
        monitor.Tick(0);
        recorder.Calls.Clear();

        monitor.Resolve(AwayResolution.Keep);

        Assert.Equal((awayStart, now, true), Assert.Single(recorder.Resolved));
        Assert.Equal(IdleState.Active, monitor.State);
        Assert.Equal(["resolve", "start"], recorder.Calls);
    }

    [Fact]
    public void DiscardResolvesWithoutKeepingAndStillReopens()
    {
        var now = T0;
        var (monitor, recorder) = New(() => now);
        monitor.Activate();
        monitor.Tick(600);
        now = T0.AddMinutes(30);
        monitor.Tick(0);

        monitor.Resolve(AwayResolution.Discard);

        Assert.False(Assert.Single(recorder.Resolved).Keeping);
        Assert.Equal(IdleState.Active, monitor.State);
    }

    /// <summary>
    /// The prompt is not modal and the app can be signed out from under it. A second answer must
    /// not produce a second bridge span.
    /// </summary>
    [Fact]
    public void ResolvingTwiceOnlyCountsOnce()
    {
        var now = T0;
        var (monitor, recorder) = New(() => now);
        monitor.Activate();
        monitor.Tick(600);
        now = T0.AddMinutes(30);
        monitor.Tick(0);

        monitor.Resolve(AwayResolution.Keep);
        monitor.Resolve(AwayResolution.Keep);

        Assert.Single(recorder.Resolved);
    }

    [Fact]
    public void TearingDownWhileAwayRecordsUnresolved()
    {
        var now = T0;
        var (monitor, recorder) = New(() => now);
        monitor.Activate();
        monitor.Tick(600);
        var awayStart = now.AddSeconds(-600);

        now = T0.AddMinutes(45);
        monitor.Deactivate();

        Assert.Equal((awayStart, now), Assert.Single(recorder.Abandoned));
        Assert.Equal(IdleState.Inactive, monitor.State);
    }

    /// <summary>
    /// Torn down after the person came back but before they answered: the window ends when they
    /// RETURNED, not when the app quit. Using "now" would silently extend the unresolved window
    /// over time the person was demonstrably at the machine.
    /// </summary>
    [Fact]
    public void TearingDownWhileAwaitingUsesTheResumeInstant()
    {
        var now = T0;
        var (monitor, recorder) = New(() => now);
        monitor.Activate();
        monitor.Tick(600);
        var awayStart = now.AddSeconds(-600);
        now = T0.AddMinutes(30);
        monitor.Tick(0);
        var resumedAt = now;

        now = T0.AddHours(2);
        monitor.Deactivate();

        Assert.Equal((awayStart, resumedAt), Assert.Single(recorder.Abandoned));
    }

    [Fact]
    public void TearingDownWhileMerelyActiveRecordsNothing()
    {
        var (monitor, recorder) = New(() => T0);
        monitor.Activate();
        recorder.Calls.Clear();

        monitor.Deactivate();

        Assert.Empty(recorder.Calls);
        Assert.Equal(IdleState.Inactive, monitor.State);
    }

    [Fact]
    public void ResumeIsANoOpUnlessAway()
    {
        var (monitor, recorder) = New(() => T0);
        monitor.Activate();
        recorder.Calls.Clear();

        monitor.Resume();

        Assert.Equal(IdleState.Active, monitor.State);
        Assert.Empty(recorder.Calls);
    }
}
